File: packages/linkedin_browser/incident.py
"""Shared LinkedIn browser incident gate and cross-lane operation lock."""

from __future__ import annotations

import fcntl
import json
import os
import uuid
from collections.abc import Iterator, Mapping, Sequence
from contextlib import contextmanager
from datetime import UTC, datetime
from enum import StrEnum
from pathlib import Path

from pydantic import BaseModel, ConfigDict, Field

from packages.linkedin_common.paths import DEFAULT_STATE_ROOT

INCIDENT_PATH_ENV = "LINKEDIN_TOOLS_INCIDENT_PATH"
BROWSER_LOCK_PATH_ENV = "LINKEDIN_TOOLS_BROWSER_LOCK_PATH"


class IncidentKind(StrEnum):
    MANUAL_PAUSE = "manual-pause"
    UNUSUAL_ACTIVITY = "unusual-activity"
    RATE_LIMIT = "rate-limit"
    LOGIN_REQUIRED = "login-required"
    CHECKPOINT = "checkpoint"
    SECURITY_CHALLENGE = "security-challenge"
    WEEKLY_LIMIT = "weekly-limit"
    NETWORK_REFUSAL = "network-refusal"


class IncidentObservation(BaseModel):
    model_config = ConfigDict(extra="forbid")

    observed_at: datetime = Field(default_factory=lambda: datetime.now(UTC))
    kind: IncidentKind
    source: str
    operation: str | None = None
    summary: str
    evidence_path: str | None = None


class LinkedInIncident(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: uuid.UUID = Field(default_factory=uuid.uuid4)
    opened_at: datetime = Field(default_factory=lambda: datetime.now(UTC))
    active: bool = True
    observations: list[IncidentObservation] = Field(default_factory=list)
    cleared_at: datetime | None = None
    clearance_reason: str | None = None
    account_access_confirmed: bool = False
    warning_cleared_confirmed: bool = False

    @property
    def latest(self) -> IncidentObservation:
        if not self.observations:
            raise ValueError("incident has no observations")
        return self.observations[-1]


class ActiveLinkedInIncidentError(RuntimeError):
    """Raised before browser work when the shared incident gate is active."""

    def __init__(self, incident: LinkedInIncident) -> None:
        self.incident = incident
        latest = incident.latest
        super().__init__(
            "LinkedIn browser work is paused by active incident "
            f"{incident.id} ({latest.kind.value}: {latest.summary})"
        )


class LinkedInIncidentDetectedError(RuntimeError):
    """Raised after a browser operation records a new fatal incident."""

    def __init__(self, incident: LinkedInIncident) -> None:
        self.incident = incident
        latest = incident.latest
        super().__init__(
            "LinkedIn browser operation stopped and opened incident "
            f"{incident.id} ({latest.kind.value}: {latest.summary})"
        )


def default_incident_path() -> Path:
    configured = os.environ.get(INCIDENT_PATH_ENV)
    return Path(configured) if configured else DEFAULT_STATE_ROOT / "linkedin-incident.json"


def default_browser_lock_path() -> Path:
    configured = os.environ.get(BROWSER_LOCK_PATH_ENV)
    return Path(configured) if configured else DEFAULT_STATE_ROOT / "browser-operation.lock"


def load_incident(path: Path | None = None) -> LinkedInIncident | None:
    target = path or default_incident_path()
    if not target.exists():
        return None
    return LinkedInIncident.model_validate_json(target.read_text(encoding="utf-8"))


def active_incident(path: Path | None = None) -> LinkedInIncident | None:
    incident = load_incident(path)
    return incident if incident is not None and incident.active else None


def assert_no_active_incident(path: Path | None = None) -> None:
    incident = active_incident(path)
    if incident is not None:
        raise ActiveLinkedInIncidentError(incident)


def open_incident(
    *,
    kind: IncidentKind,
    source: str,
    summary: str,
    operation: str | None = None,
    evidence_path: str | None = None,
    path: Path | None = None,
) -> LinkedInIncident:
    target = path or default_incident_path()
    incident = active_incident(target) or LinkedInIncident()
    incident.observations.append(
        IncidentObservation(
            kind=kind,
            source=source,
            operation=operation,
            summary=summary,
            evidence_path=evidence_path,
        )
    )
    _write_incident(target, incident)
    return incident


def clear_incident(
    *,
    reason: str,
    account_access_confirmed: bool,
    warning_cleared_confirmed: bool,
    path: Path | None = None,
) -> LinkedInIncident:
    target = path or default_incident_path()
    incident = active_incident(target)
    if incident is None:
        raise RuntimeError("no active LinkedIn incident exists")
    if not account_access_confirmed or not warning_cleared_confirmed:
        raise RuntimeError(
            "clearing the LinkedIn incident requires confirmed account access "
            "and confirmed warning clearance"
        )
    if not reason.strip():
        raise RuntimeError("clearing the LinkedIn incident requires a reason")
    incident.active = False
    incident.cleared_at = datetime.now(UTC)
    incident.clearance_reason = reason.strip()
    incident.account_access_confirmed = True
    incident.warning_cleared_confirmed = True
    _write_incident(target, incident)
    return incident


@contextmanager
def browser_operation_lock(
    operation: str,
    *,
    path: Path | None = None,
) -> Iterator[None]:
    target = path or default_browser_lock_path()
    target.parent.mkdir(parents=True, exist_ok=True)
    handle = target.open("a+", encoding="utf-8")
    try:
        try:
            fcntl.flock(handle.fileno(), fcntl.LOCK_EX | fcntl.LOCK_NB)
        except BlockingIOError as exc:
            handle.seek(0)
            owner = handle.read().strip() or "owner details unavailable"
            raise RuntimeError(
                f"another LinkedIn browser operation is active: {owner}"
            ) from exc
        handle.seek(0)
        handle.truncate()
        handle.write(
            json.dumps(
                {
                    "pid": os.getpid(),
                    "operation": operation,
                    "started_at": datetime.now(UTC).isoformat(),
                },
                sort_keys=True,
            )
        )
        handle.flush()
        yield
    finally:
        try:
            fcntl.flock(handle.fileno(), fcntl.LOCK_UN)
        finally:
            handle.close()


def detect_fatal_incident(value: object) -> tuple[IncidentKind, str] | None:
    strings = _flatten_scalar_strings(value)
    normalized = "\n".join(strings).casefold()
    rules: Sequence[tuple[IncidentKind, tuple[str, ...], str]] = (
        (
            IncidentKind.UNUSUAL_ACTIVITY,
            ("we noticed some unusual activity on your account",),
            "LinkedIn unusual-activity warning",
        ),
        (
            IncidentKind.RATE_LIMIT,
            (
                "http 429",
                "status 429",
                "httpstatus\n429",
                "http_status\n429",
                "returned http 429",
                "profile api rate limited",
                "sales-nav-profile-api-rate-limited",
            ),
            "LinkedIn returned HTTP 429",
        ),
        (
            IncidentKind.CHECKPOINT,
            ("checkpoint present", "/checkpoint",),
            "LinkedIn checkpoint present",
        ),
        (
            IncidentKind.SECURITY_CHALLENGE,
            ("security verification present", "security challenge"),
            "LinkedIn security verification present",
        ),
        (
            IncidentKind.LOGIN_REQUIRED,
            ("login required", "/uas/login", "linkedin.com/login"),
            "LinkedIn login required",
        ),
        (
            IncidentKind.WEEKLY_LIMIT,
            ("weekly limit",),
            "LinkedIn weekly limit reached",
        ),
        (
            IncidentKind.NETWORK_REFUSAL,
            (
                "net::err_name_not_resolved",
                "network refusal",
                "connection refused",
            ),
            "LinkedIn browser network request failed",
        ),
    )
    for kind, needles, summary in rules:
        if any(needle in normalized for needle in needles):
            return kind, summary
    return None


def inspect_artifact_for_fatal_incident(path: Path) -> tuple[IncidentKind, str] | None:
    if not path.exists():
        return None
    try:
        payload = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, ValueError):
        return None
    return detect_fatal_incident(payload)


def _flatten_scalar_strings(value: object) -> list[str]:
    result: list[str] = []

    def visit(item: object) -> None:
        if isinstance(item, str):
            result.append(item)
            return
        if isinstance(item, int | float | bool):
            result.append(str(item))
            return
        if isinstance(item, Mapping):
            for key, child in item.items():
                if isinstance(key, str):
                    result.append(key)
                visit(child)
            return
        if isinstance(item, Sequence) and not isinstance(item, (str, bytes, bytearray)):
            for child in item:
                visit(child)

    visit(value)
    return result


def _write_incident(path: Path, incident: LinkedInIncident) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    temporary = path.with_suffix(path.suffix + ".tmp")
    temporary.write_text(
        json.dumps(incident.model_dump(mode="json"), indent=2) + "\n",
        encoding="utf-8",
    )
    temporary.replace(path)

File: src/browser/incident.ts
import { mkdir, readFile, rename, unlink, writeFile } from "node:fs/promises";
import { join } from "node:path";

export const INCIDENT_FILE_NAME = "linkedin-incident.json";
export const INCIDENT_ACTIVE_CODE = "INCIDENT_ACTIVE";
export const INCIDENT_ACTIVE_EXIT_CODE = 7;

export type IncidentKind =
  | "weekly_limit"
  | "rate_limit"
  | "unusual_activity"
  | "checkpoint"
  | "security_verification"
  | "login_required"
  | "unknown";

export type LinkedInIncident = {
  readonly kind: IncidentKind;
  readonly reason: string;
  readonly detail: string;
  readonly opened_at: string;
  readonly active: boolean;
  readonly cleared_at?: string;
  readonly clearance_reason?: string;
  readonly account_access_confirmed?: boolean;
  readonly warning_cleared_confirmed?: boolean;
};

export class ActiveIncidentError extends Error {
  readonly code = INCIDENT_ACTIVE_CODE;
  readonly exitCode = INCIDENT_ACTIVE_EXIT_CODE;
  readonly incident: LinkedInIncident;

  constructor(incident: LinkedInIncident) {
    const summary = `${incident.kind}: ${incident.reason}`;
    super(`LinkedIn browser work is paused by active incident (${summary})`);
    this.name = "ActiveIncidentError";
    this.incident = incident;
  }
}

export class IncidentDetectedError extends Error {
  readonly code = INCIDENT_ACTIVE_CODE;
  readonly exitCode = INCIDENT_ACTIVE_EXIT_CODE;
  readonly incident: LinkedInIncident;

  constructor(incident: LinkedInIncident) {
    const summary = `${incident.kind}: ${incident.reason}`;
    super(`LinkedIn browser operation stopped and opened incident (${summary})`);
    this.name = "IncidentDetectedError";
    this.incident = incident;
  }
}

const FATAL_RULES: readonly {
  readonly kind: IncidentKind;
  readonly needles: readonly string[];
  readonly reason: string;
}[] = [
  {
    kind: "unusual_activity",
    needles: ["unusual activity", "we noticed some unusual activity on your account"],
    reason: "LinkedIn unusual-activity warning",
  },
  {
    kind: "rate_limit",
    needles: ["http 429", "status 429", "returned http 429", "429"],
    reason: "LinkedIn returned HTTP 429",
  },
  {
    kind: "checkpoint",
    needles: ["checkpoint present", "/checkpoint", "checkpoint"],
    reason: "LinkedIn checkpoint present",
  },
  {
    kind: "security_verification",
    needles: ["security verification", "security challenge", "quick security check"],
    reason: "LinkedIn security verification present",
  },
  {
    kind: "login_required",
    needles: ["session_key", "authwall", "login required", "/uas/login", "linkedin.com/login"],
    reason: "LinkedIn login required",
  },
  {
    kind: "weekly_limit",
    needles: [
      "weekly limit",
      "weekly invitation limit",
      "weekly connection limit",
      "you've reached the weekly",
      "you’ve reached the weekly",
    ],
    reason: "LinkedIn weekly limit reached",
  },
];

/** Ordinary automation blockers must never open the shared incident gate. */
const ORDINARY_BLOCKER_MARKERS = [
  "WRONG_PAGE",
  "WORKFLOW_PAGE_MISSING",
  "WORKFLOW_PAGE_AMBIGUOUS",
  "SELECTOR_CONTRACT",
  "ALREADY_PENDING",
  "EMAIL_REQUIRED",
  "MISSING_MORE_ACTIONS",
  "MISSING_CONNECT_MENU",
  "MISSING_SEND",
  "DISABLED_SEND",
  "CANDIDATE_ABSENT",
  "ROW_LOAD_TIMEOUT",
  "NO_ROWS",
  "STALLED_NAVIGATION",
  "SOURCE_EXHAUSTED",
  "SOURCE_MISMATCH",
  "UNCLEAR_CONFIRMATION",
  "PREPARATION_MISMATCH",
  "PREPARATION_STALE",
  "COMMIT_SEND_UNCERTAIN",
  "EVIDENCE_CORRUPT",
  "EVIDENCE_FINALIZATION",
  "EVIDENCE_HANDOFF",
] as const;

export function incidentPath(stateDir: string): string {
  return join(stateDir, INCIDENT_FILE_NAME);
}

export async function loadIncident(stateDir: string): Promise<LinkedInIncident | null> {
  const path = incidentPath(stateDir);
  let text: string;
  try {
    text = await readFile(path, "utf8");
  } catch (error) {
    if (typeof error === "object" && error !== null && "code" in error && error.code === "ENOENT") {
      return null;
    }
    throw error;
  }
  try {
    const parsed: unknown = JSON.parse(text);
    if (
      typeof parsed !== "object" ||
      parsed === null ||
      !("kind" in parsed) ||
      !("reason" in parsed) ||
      !("detail" in parsed) ||
      !("opened_at" in parsed) ||
      !("active" in parsed) ||
      typeof parsed.kind !== "string" ||
      typeof parsed.reason !== "string" ||
      typeof parsed.detail !== "string" ||
      typeof parsed.opened_at !== "string" ||
      typeof parsed.active !== "boolean"
    ) {
      return null;
    }
    return parsed as LinkedInIncident;
  } catch {
    return null;
  }
}

export async function activeIncident(stateDir: string): Promise<LinkedInIncident | null> {
  const incident = await loadIncident(stateDir);
  return incident !== null && incident.active ? incident : null;
}

export async function assertNoActiveIncident(stateDir: string): Promise<void> {
  const incident = await activeIncident(stateDir);
  if (incident !== null) throw new ActiveIncidentError(incident);
}

export async function openIncident(
  stateDir: string,
  input: {
    readonly kind: IncidentKind;
    readonly reason: string;
    readonly detail: string;
    readonly openedAt?: string;
  },
): Promise<LinkedInIncident> {
  const existing = await activeIncident(stateDir);
  if (existing !== null) return existing;
  const incident: LinkedInIncident = {
    kind: input.kind,
    reason: input.reason,
    detail: input.detail.slice(0, 2000),
    opened_at: input.openedAt ?? new Date().toISOString(),
    active: true,
  };
  await writeIncident(stateDir, incident);
  return incident;
}

export async function clearIncident(
  stateDir: string,
  input: {
    readonly reason: string;
    readonly accountAccessConfirmed: boolean;
    readonly warningClearedConfirmed: boolean;
    readonly clearedAt?: string;
  },
): Promise<LinkedInIncident> {
  const incident = await activeIncident(stateDir);
  if (incident === null) {
    throw new Error("no active LinkedIn incident exists");
  }
  if (!input.accountAccessConfirmed || !input.warningClearedConfirmed) {
    throw new Error(
      "clearing the LinkedIn incident requires confirmed account access and confirmed warning clearance",
    );
  }
  const reason = input.reason.trim();
  if (reason.length === 0) {
    throw new Error("clearing the LinkedIn incident requires a reason");
  }
  const cleared: LinkedInIncident = {
    ...incident,
    active: false,
    cleared_at: input.clearedAt ?? new Date().toISOString(),
    clearance_reason: reason,
    account_access_confirmed: true,
    warning_cleared_confirmed: true,
  };
  await writeIncident(stateDir, cleared);
  return cleared;
}

export function detectFatalIncident(
  value: unknown,
): { readonly kind: IncidentKind; readonly reason: string; readonly detail: string } | null {
  const strings = flattenScalarStrings(value);
  if (strings.length === 0) return null;
  const joined = strings.join("\n");
  const normalized = joined.toLowerCase();
  const matched = FATAL_RULES.find((rule) =>
    rule.needles.some((needle) => normalized.includes(needle.toLowerCase())),
  );
  if (matched === undefined) return null;
  // Ordinary automation blockers alone never open the shared gate; a true fatal
  // needle still wins when both appear in the same failure text.
  const primary = matched.needles[0] ?? matched.reason;
  const index = normalized.indexOf(primary.toLowerCase());
  const detail =
    index < 0
      ? joined.slice(0, 400)
      : joined
          .slice(Math.max(0, index - 80), Math.min(joined.length, index + primary.length + 200))
          .replace(/\s+/g, " ")
          .trim();
  return { kind: matched.kind, reason: matched.reason, detail };
}

/**
 * Scan failure text and open the shared gate when a fatal LinkedIn signal is present.
 * Ordinary blockers alone never open an incident.
 */
export async function maybeOpenIncidentFromFailure(
  stateDir: string,
  value: unknown,
): Promise<LinkedInIncident | null> {
  if (
    typeof value === "string" &&
    ORDINARY_BLOCKER_MARKERS.some((marker) => value.includes(marker))
  ) {
    const hasFatal = FATAL_RULES.some((rule) =>
      rule.needles.some((needle) => value.toLowerCase().includes(needle.toLowerCase())),
    );
    if (!hasFatal) return null;
  }
  const detected = detectFatalIncident(value);
  if (detected === null) return null;
  return openIncident(stateDir, detected);
}

async function writeIncident(stateDir: string, incident: LinkedInIncident): Promise<void> {
  await mkdir(stateDir, { recursive: true, mode: 0o700 });
  const destination = incidentPath(stateDir);
  const temporary = join(stateDir, `.${INCIDENT_FILE_NAME}.${process.pid}.tmp`);
  const body = `${JSON.stringify(incident, null, 2)}\n`;
  try {
    await writeFile(temporary, body, { encoding: "utf8", mode: 0o600 });
    await rename(temporary, destination);
  } catch (error) {
    try {
      await unlink(temporary);
    } catch {
      // best-effort cleanup
    }
    throw error;
  }
}

function flattenScalarStrings(value: unknown): string[] {
  const result: string[] = [];
  const visit = (item: unknown): void => {
    if (typeof item === "string") {
      result.push(item);
      return;
    }
    if (typeof item === "number" || typeof item === "boolean") {
      result.push(String(item));
      return;
    }
    if (item === null || item === undefined) return;
    if (Array.isArray(item)) {
      for (const child of item) visit(child);
      return;
    }
    if (typeof item === "object") {
      for (const [key, child] of Object.entries(item)) {
        result.push(key);
        visit(child);
      }
    }
  };
  visit(value);
  return result;
}
